modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/file/AbstractFileIO.java

```diff
@@ -0,0 +1,129 @@
+package org.apache.ignite.internal.processors.cache.persistence.file;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import org.apache.ignite.internal.util.typedef.internal.U;
+
+/**
+ *
+ */
+public abstract class AbstractFileIO implements FileIO {
+    /** Max io timeout milliseconds. */
+    private static final int MAX_IO_TIMEOUT_MS = 2000;
+
+    /**
+     *
+     */
+    private interface IOOperation {
+        /**
+         * @param offs Offset.
+         *
+         * @return Number of bytes operated.
+         */
+        public int run(int offs) throws IOException;
+    }
+
+    /**
+     * @param operation IO operation.
+     *
+     * @param num Number of bytes to operate.
+     */
+    private int fully(IOOperation operation, int num, boolean write) throws IOException {
+        if (num > 0) {
+            long time = 0;
+
+            for (int i = 0; i < num; ) {
+                int n = operation.run(i);
+
+                if (n > 0) {
+                    i += n;
+                    time = 0;
+                }
+                else if (n == 0) {
+                    if (time == 0)
+                        time = U.currentTimeMillis();
+                    else if ((U.currentTimeMillis() - time) >= MAX_IO_TIMEOUT_MS)
+                        throw new IOException(write && position() == size() ? "Failed to extend file." :
```

Comment:
Change message for this exception. Failed to extend failed was in WAL. This method is generic)